### PR TITLE
refactor(ts-estree): add types to converter

### PR DESCRIPTION
**Pull request by armano2**
This is prototype of converter with types, with small improvement to performance (this is not goal of this PR).

Changes:
- type checking of AST structure
- reduces amount of calls for location data (location is calculated once per node)
- provides easier way to support range compiler flag (its not supported yet)
- remove old/unused estree definition files
- migrates ParserOptions from `temp-types-based-on-js-source.ts` to `parser-options.ts`

<details><summary>...</summary>

**before**
Complex File: `ts-estree x 238 ops/sec ±2.14% (86 runs sampled)`
Simple File: `ts-estree x 10,706 ops/sec ±3.19% (87 runs sampled)`

**after**
Complex File: `ts-estree x 262 ops/sec ±1.78% (84 runs sampled)`
Simple File: `ts-estree x 11,049 ops/sec ±3.26% (80 runs sampled)`

</details>

### CODE REVIEW

**Comment by bradzacher** [comment]
we should really raise a separate PR to remove all of the JSDoc types from the codebase...

**Comment by armano2** [comment]
both PR adds same file as ast structure, and they are not blocking each other.

File: packages/typescript-estree/src/convert.ts [hunk 1]
```diff
@@ -297,83 +327,106 @@ export default function convert(config: ConvertConfig): ESTreeNode | null {
...
 
   /**
    * Applies the given TS modifiers to the given result object.
-   * @param {ts.ModifiersArray} modifiers original ts.Nodes from the node.modifiers array
-   * @returns {void} (the current result object will be mutated)
+   * @param result
+   * @param modifiers original ts.Nodes from the node.modifiers array
+   * @returns the current result object will be mutated
+   * @deprecated
+   * TODO: remove this
```
Comment (JamesHenry) [inline packages/typescript-estree/src/convert.ts]: Can you expand on this comment?

```
TODO: remove this
```

Is an action that could be completed at any time, I'm guessing something has to happen before that is appropriate? Right now it is unclear

File: packages/typescript-estree/src/convert.ts [hunk 2]
```diff
@@ -417,2091 +470,2023 @@ export default function convert(config: ConvertConfig): ESTreeNode | null {
...
+          local: this.convertChild(node.name)
         });
-      }
-      break;
 
-    case SyntaxKind.ExportSpecifier:
-      Object.assign(result, {
-        type: AST_NODE_TYPES.ExportSpecifier,
-        local: convertChild(node.propertyName || node.name),
-        exported: convertChild(n
```
Comment (JamesHenry) [inline packages/typescript-estree/src/convert.ts]: Again, `TODO` comment is unclear, please elaborate

Comment (armano2) [reply packages/typescript-estree/src/convert.ts]: ohh, i missed this one, i will remove it

Comment (armano2) [reply packages/typescript-estree/src/convert.ts]: while developing i'm adding a lot of todos to remind myself that i have to fix stuff

File: packages/typescript-estree/src/convert.ts (hunk 1 above)
Comment (armano2) [reply packages/typescript-estree/src/convert.ts]: in all nodes except 2 we are handing unsupported modifers in diferent way than for `es.TSEnumDeclaration | es.TSModuleDeclaration`, i should write appropriate message here

Comment (armano2) [reply packages/typescript-estree/src/convert.ts]:
```
* @deprecated This function adds not standardized `modifiers` property in nodes
```

**Comment by JamesHenry** [comment]
I manually ran some real-world "benchmarks" by building from this PR and including it in a build of Prettier and running it on a real codebase (angular.io documentation website) and can confirm there are not any performance regressions/issues

**Comment by JamesHenry** [comment]
Thanks for all the work on this!

**Pull request merged**